ablySpec/Utilities.swift

```diff
@@ -0,0 +1,137 @@
+                    expect(receivedFoo2).to(equal(123))
+                    expect(receivedBar).to(beNil())
+                    expect(receivedFooOrBar).to(equal(123))
+                    expect(receivedAll).to(equal(123))
+                    
+                    eventEmitter.emit("bar", with:456)
+                    
+                    expect(receivedFoo1).to(equal(123))
+                    expect(receivedFoo2).to(equal(123))
+                    expect(receivedBar).to(equal(456))
+                    expect(receivedFooOrBar).to(equal(456))
+                    expect(receivedAll).to(equal(456))
+                    
+                    eventEmitter.emit("qux", with:789)
+                    
+                    expect(receivedFooOrBar).to(equal(456))
+                    expect(receivedAll).to(equal(789))
+                }
+                
+                it("should only call once listeners once for its event") {
+                    eventEmitter.emit("foo", with: 123)
+
+                    expect(receivedBarOnce).to(beNil())
+                    expect(receivedAllOnce).to(equal(123))
+                    
+                    eventEmitter.emit("bar", with: 456)
+                    
+                    expect(receivedBarOnce).to(equal(456))
+                    expect(receivedAllOnce).to(equal(123))
+
+                    eventEmitter.emit("bar", with: 789)
+                    
+                    expect(receivedBarOnce).to(equal(456))
+                    expect(receivedAllOnce).to(equal(123))
+                }
+                
+                it("should stop receiving events when calling off") {
+                    eventEmitter.offAll(listenerFoo1!)
+                    eventEmitter.emit("foo", with: 123)
+                    
+                    expect(receivedFoo1).to(beNil())
+                    expect(receivedFoo2).to(equal(123))
+                    expect(receivedFooOrBar).to(equal(123))
+                    expect(receivedAll).to(equal(123))
+                    
+                    eventEmitter.off("foo", listener: listenerFooOrBar!)
+                    eventEmitter.emit("foo", with: 456)
+                    
+                    expect(receivedFoo2).to(equal(456))
+                    expect(receivedFooOrBar).to(equal(123))
+                    
+                    eventEmitter.emit("bar", with: 789)
+                    
+                    expect(receivedFooOrBar).to(equal(789))
+                    
+                    eventEmitter.off("foo", listener: listenerAll!)
+                    eventEmitter.emit("foo", with: 111)
+                    
+                    expect(receivedAll).to(equal(789))
+                    
+                    eventEmitter.emit("bar", with: 222)
+                    
+                    expect(receivedAll).to(equal(222))
+                }
+
+                it("should receive only once after calling onAll and then once") {
```

Comment:
Hmm, OK, that changes things. It's simpler to implement also. I prefer the behavior I implemented, as I think set-like on/off operations are more robust and more coherent with the API than list-like behavior (as you just noticed yourself), but well.